go/vt/tabletmanager/rpc_replication.go

```diff
@@ -555,14 +547,63 @@ func isMasterEligible(tabletType topodatapb.TabletType) bool {
 	return false
 }
 
-func (agent *ActionAgent) enableSemiSync(master bool) error {
+func (agent *ActionAgent) setSemiSync(tabletType topodatapb.TabletType) error {
+	if !*enableSemiSync {
+		// Semi-sync handling is not enabled.
+		return nil
+	}
+
 	// Only enable if we're eligible for becoming master (REPLICA type).
 	// Ineligible slaves (RDONLY) shouldn't ACK because we'll never promote them.
-	if !isMasterEligible(agent.Tablet().Type) {
-		return nil
+	if !isMasterEligible(tabletType) {
+		return agent.MysqlDaemon.SetSemiSyncEnabled(false, false)
 	}
 
 	// Always enable slave-side since it doesn't hurt to keep it on for a master.
 	// The master-side needs to be off for a slave, or else it will get stuck.
-	return agent.MysqlDaemon.SetSemiSyncEnabled(master, true)
+	return agent.MysqlDaemon.SetSemiSyncEnabled(tabletType == topodatapb.TabletType_MASTER, true)
+}
+
+func (agent *ActionAgent) setSemiSyncAndFixReplication(tabletType topodatapb.TabletType) error {
+	if !*enableSemiSync {
+		// Semi-sync handling is not enabled.
+		return nil
+	}
+
+	if tabletType == topodatapb.TabletType_MASTER {
+		// Master is special. It is always handled at the
+		// right time by the reparent operations, it doesn't
+		// need to be fixed.
+		return nil
+	}
+
+	if err := agent.setSemiSync(tabletType); err != nil {
+		return fmt.Errorf("failed to setSemiSync(%v): %v", tabletType, err)
+	}
+
+	// If replication is running, but the status is wrong,
+	// restart replication.
+	if _, err := agent.MysqlDaemon.SlaveStatus(); err != nil {
+		// Replication is not running, nothing to do.
+		return nil
+	}
+
+	shouldAck := isMasterEligible(tabletType)
+	acking, err := agent.MysqlDaemon.SemiSyncSlaveStatus()
+	if err != nil {
+		return fmt.Errorf("failed to get SemiSyncSlaveStatus: %v", err)
+	}
+	if shouldAck == acking {
+		return nil
+	}
+
+	// We need to restart replication
+	log.Infof("Restarting replication for semi-sync flag change to take effect from %v to %v", acking, shouldAck)
+	if err := mysqlctl.StopSlave(agent.MysqlDaemon, agent.hookExtraEnv()); err != nil {
+		return fmt.Errorf("failed to StopSlave: %v", err)
+	}
+	if err := mysqlctl.StartSlave(agent.MysqlDaemon, agent.hookExtraEnv()); err != nil {
```

Comment:
BTW, should we account here for a case when someone executed StopSlave first and then ChangeType, i.e. they don't want StartSlave to be automatically executed on any ChangeType? Besides if I'm not mistaken in such case StopSlave above will fail because the slave is already stopped.